### ISSUE TITLE
Overridden interface default methods are not considered overridden

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#978]
## Overview

This is a follow-up to #976.

Whether or not the bug described in _this_ issue existed at a previous point in time is currently unknown (simply because it has not been researched).

In any case, we now have a situation where an _overridden_ interface default method is not considered to be overridden by the `findMethods()` search algorithm in `ReflectionUtils`.

Consequently, if we execute all tests in the following `GenericTestCaseWithOverriddenDefaultMethod` JUnit Jupiter based test class, we see that 3 tests are executed instead of 2. Specifically, we expect that only `test(Long)` and `test(Double)` will be executed; however, `test(Number)` is additionally _discovered_ resulting in `test(Long)` being executed twice.

```java
@ExtendWith({ LongParameterResolver.class, DoubleParameterResolver.class })
interface GenericTestInterface<N extends Number> {

	@Test
	default void test(N number) {
	}
}

class GenericTestCaseWithOverriddenDefaultMethod implements GenericTestInterface<Long> {

	@Test
	@Override
	public void test(Long number) {
	}

	@Test
	void test(Double number) {
	}
}
```

## Analysis

The recent series of [Whac-A-Mole](https://en.wikipedia.org/wiki/Whac-A-Mole) bug fixes in this area leads to me believe that the only _real_ solution to this set of problems is one that takes into account the actual generic and parameterized type information instead of relying on a best-guess approach involving method sub-signatures and blind checks for the presence of generics.

Depending on how we _take into account the actual generic and parameterized type information_, we might potentially wish to reverse a subset of changes in fae5b6a4f001e7427fba13b13fb844074d5ff099 that explicitly treat default methods as locally declared methods (while leaving the other changes in that commit in tact).

## Further Considerations

This issue currently only focuses on overridden interface default methods; however, it may well be the case that our `findMethods()` search algorithm contains other bugs related to generics.

## Related Issues

- #333 
- #976 

## Deliverables

- [x] Ensure that overridden interface default methods are not returned in the results of `ReflectionUtils.findMethods()`.
- [x] Investigate possible additional bugs hinted at in _Further Considerations_.
  - to be addressed in #981

**Comment by sbrannen** [comment]
Added _Further Considerations_ section.

**Comment by jbduncan** [comment]
My understanding of the interplay between generics and reflection is a bit limited, but I would think that https://github.com/jhalterman/typetools or something similar would be helpful here.

**Comment by sbrannen** [comment]
> My understanding of the interplay between generics and reflection is a bit limited, but I would think that https://github.com/jhalterman/typetools or something similar would be helpful here.

Yeah, that might prove useful, but we are initially attempting to solve the issue without including an additional third-party library.

**Comment by sormuras** [comment]
Found a solution that fixes the "is method overridden problem" using global view on all visible method. If they are generic or not does not matter here. The compiler and runtime already sorted that out.

Needs some performance considerations though.

**Comment by sbrannen** [comment]
Fixed by @sormuras in commit 9c9df1d15b8dbcf62c27fe0bf329ce73118dbd35.

Additional testing will be addressed in #981.
